### PR TITLE
Move macaron to chi

### PR DESCRIPTION
**Pull request by lunny**
Use [chi](https://github.com/go-chi/chi) instead of the forked [macaron](https://gitea.com/macaron/macaron). Since macaron and chi have conflicts with session share, this big PR becomes a have-to thing. According my previous idea, we can replace macaron step by step but I'm wrong. :( Below is a list of big changes on this PR.

- [x] Define `context.ResponseWriter` interface with an implementation `context.Response`.
- [x] Use chi instead of macaron, and also a customize `Route` to wrap chi so that the router usage is similar as before.
- [x] Create different routers for `web`, `api`, `internal` and `install` so that the codes will be more clear and no magic .
- [x] Use https://github.com/unrolled/render instead of macaron's internal render
- [x] Use https://github.com/NYTimes/gziphandler instead of https://gitea.com/macaron/gzip
- [x] Use https://gitea.com/go-chi/session which is a modified version of https://gitea.com/macaron/session and removed `nodb` support since it will not be maintained. **BREAK**
- [x] Use https://gitea.com/go-chi/captcha which is a modified version of https://gitea.com/macaron/captcha
- [x] Use https://gitea.com/go-chi/cache which is a modified version of https://gitea.com/macaron/cache
- [x] Use https://gitea.com/go-chi/binding which is a modified version of https://gitea.com/macaron/binding
- [x] Use https://github.com/go-chi/cors instead of https://gitea.com/macaron/cors
- [x] Dropped https://gitea.com/macaron/i18n and make a new one in `code.gitea.io/gitea/modules/translation`
- [x] Move validation form structs from `code.gitea.io/gitea/modules/auth` to `code.gitea.io/gitea/modules/forms` to avoid dependency cycle.
- [x] Removed macaron log service because it's not need any more. **BREAK**
- [x] All form structs have to be get by `web.GetForm(ctx)` in the route function but not as a function parameter on routes definition.
- [x] Move Git HTTP protocol implementation to use routers directly.
- [x] Fix the problem that chi routes don't support trailing slash but macaron did.
- [x] `/api/v1/swagger` now will be redirect to `/api/swagger` but not render directly so that `APIContext` will not create a html render. 

Notices:
- Chi router don't support request with trailing slash
- Integration test `TestUserHeatmap` maybe mysql version related. It's failed on my macOS(mysql 5.7.29 installed via brew) but succeed on CI.

UPDATE: **2021.01.27**
- [x] environment variables `HOST`, `PORT` and `MACARON_ENV` has been removed. **BREAK**

### CODE REVIEW

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Review by jace788 — APPROVED** [review]

**Comment by 6543** [comment]
I'd like to mention I got a not reproducable panic one time at **test-sqlite**: https://pastebin.com/kj4hvAf8
It should be at least recorded ... here

File: modules/context/form.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	if len(defaults) > 0 && err != nil {
+		return defaults[0]
+	}
+	return uint(v)
+}
+
+// MustUint32 returns request form as uint32 with default
+func (f *Forms) MustUint32(key string, defaults ...uint32) uint32 {
+	v, err := strconv.ParseUint((*http.Request)(f).FormValue(key), 10, 32)
+	if len(defaults) > 0 && err != nil {
```
Comment (6543) [inline modules/context/form.go:187]: just a thought is `x != 0` faster or equal to `x > 0`

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -500,32 +503,66 @@ func mustEnableIssuesOrPulls(ctx *context.APIContext) {
...
+		SignInRequired: setting.Service.RequireSignInView,
+	}))
 
-	m.Group("/v1", func() {
+	m.Group("", func() {
 		// Miscellaneous
 		if setting.API.EnableSwagger {
-			m.Get("/swagger", misc.Swagger)
+			m.Get("/swagger", func(ctx *context.APIContext) {
+				ctx.Redirect(setting.AppURL + "/api/swagger")
```
Comment (6543) [inline routers/api/v1/api.go]:
```suggestion
				ctx.Redirect(setting.AppURL + "api/swagger")
```

File: routers/routes/base.go [hunk 3]
```diff
@@ -78,7 +76,7 @@ func setupAccessLogger(c chi.Router) {
 				log.Error("Could not set up macaron access logger: %v", err.Error())
```
Comment (6543) [inline routers/routes/base.go]:
```suggestion
				log.Error("Could not set up chi access logger: %v", err.Error())
```
line 71 is the same

**Review by 6543 — COMMENTED** [review]
some nits ...

File: routers/install.go [hunk 4]
```diff
@@ -33,17 +38,40 @@ const (
...
+		if setting.InstallLock {
+			resp.Header().Add("Refresh", "1; url="+setting.AppURL+"user/login")
+			_ = rnd.HTML(resp, 200, string(tplPostInstall), nil)
+			return
+		}
+		var locale = middlewares.Locale(resp, req)
+		var startTime = time.Now()
+		var ctx = context.Context{
+			Resp: context.NewResponse(resp),
+			//csrf:    x,
```
Comment (6543) [inline routers/install.go]: was it there bevore? is it a TODO?

File: routers/api/v1/api.go [hunk 5]
```diff
@@ -500,32 +503,66 @@ func mustEnableIssuesOrPulls(ctx *context.APIContext) {
...
+		Maxlifetime:    setting.SessionConfig.Maxlifetime,
+		Secure:         setting.SessionConfig.Secure,
+		Domain:         setting.SessionConfig.Domain,
+	}))
+	m.Use(securityHeaders())
+	if setting.CORSConfig.Enabled {
+		m.Use(cors.Handler(cors.Options{
+			//Scheme:           setting.CORSConfig.Scheme,
+			AllowedOrigins: setting.CORSConfig.AllowDomain,
+			//setting.CORSConfig.AllowSubdomain
```
Comment (6543) [inline routers/api/v1/api.go]: is there a reason AllowSubdomain and Scheme got commented out?

File: modules/context/form.go (hunk 1 above)
Comment (6543) [reply modules/context/form.go:187]: ok looks like this is irellevant:
```
goos: linux
goarch: amd64
BenchmarkGt-24      	1000000000	         0.251 ns/op
BenchmarkUnEQ-24    	1000000000	         0.251 ns/op
PASS
ok  	_/tmp/a/benchmark	0.558s
```

File: modules/context/auth.go [hunk 6]
```diff
@@ -108,19 +77,87 @@ func Toggle(options *ToggleOptions) macaron.Handler {
...
+	return func(ctx *APIContext) {
+		// Check prohibit login users.
+		if ctx.IsSigned {
+			if !ctx.User.IsActive && setting.Service.RegisterEmailConfirm {
+				ctx.Data["Title"] = ctx.Tr("auth.active_your_account")
+				ctx.JSON(403, map[string]string{
+					"message": "This account is not activated.",
+				})
+				return
+			} else if !ctx.User.IsActive || ctx.User.ProhibitLogin {
```
Comment (a1012112796) [inline modules/context/auth.go]: the ``else`` looks not needed.

File: modules/context/context.go [hunk 7]
```diff
@@ -249,66 +314,332 @@ func (ctx *Context) ServeContent(name string, r io.ReadSeeker, params ...interfa
...
+					Values:     vals,
+					ErrorMsg:   vals.Get("error"),
+					SuccessMsg: vals.Get("success"),
+					InfoMsg:    vals.Get("info"),
+					WarningMsg: vals.Get("warning"),
+				}
+
+				/*t, _ := strconv.ParseInt(f.Get("time"), 10, 64)
+				now := time.Now().Unix()
+				if now-t >= 0 && now-t < 3600 {*/
```
Comment (a1012112796) [inline modules/context/context.go]: please clean these not usefull code.

**Comment by lunny** [comment]
@6543 @a1012112796 All done.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
I think we should refactor some things and have a second look at all the context things - but this pull is big enough ... and works fine :)

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
.

**Pull request merged**